core/trace/src/main/scala/org/typelevel/otel4s/trace/TraceState.scala

```diff
@@ -0,0 +1,238 @@
+    * map contains no mapping for the key.
+    *
+    * @param key
+    *   with which the specified value is to be associated
+    */
+  def get(key: String): Option[String]
+
+  /** Adds or updates the entry that has the given `key` if it is present. If
+    * either the key or the value is invalid, the entry will not be added.
+    *
+    * @param key
+    *   the key for the entry to be added. Key is an opaque string up to 256
+    *   characters printable. It MUST begin with a lowercase letter, and can
+    *   only contain lowercase letters a-z, digits 0-9, underscores _, dashes -,
+    *   asterisks *, and forward slashes /. For multi-tenant vendor scenarios,
+    *   an at sign (@) can be used to prefix the vendor name. The tenant id
+    *   (before the '@') is limited to 240 characters and the vendor id is
+    *   limited to 13 characters. If in the multi-tenant vendor format, then the
+    *   first character may additionally be numeric.
+    *
+    * @param value
+    *   the value for the entry to be added. Value is opaque string up to 256
+    *   characters printable ASCII RFC0020 characters (i.e., the range 0x20 to
+    *   0x7E) except comma , and =.
+    *
+    * @return
+    *   a new instance of [[TraceState]] with added entry
+    */
+  def put(key: String, value: String): TraceState
+
+  /** Removes the entry that has the given `key` if it is present.
+    *
+    * @param key
+    *   the key for the entry to be removed.
+    *
+    * @return
+    *   a new instance of [[TraceState]] with removed entry
+    */
+  def remove(key: String): TraceState
+
+  /** Returns the number of entries in this state. */
+  def size: Int
+
+  /** Returns whether this state is empty, containing no entries. */
+  def isEmpty: Boolean
+
+  /** Returns a map representation of this state. */
+  def asMap: Map[String, String]
+
+  override final def hashCode(): Int =
+    Hash[TraceState].hash(this)
+
+  override final def equals(obj: Any): Boolean =
+    obj match {
+      case other: TraceState => Hash[TraceState].eqv(this, other)
+      case _                 => false
+    }
+
+  override final def toString: String =
+    Show[TraceState].show(this)
+}
+
+object TraceState {
+
+  /** An empty [[TraceState]]. */
+  val Empty: TraceState = MapBasedTraceState(Map.empty)
+
+  implicit val traceStateHash: Hash[TraceState] =
+    Hash.by(_.asMap)
+
+  implicit val traceStateShow: Show[TraceState] =
+    Show.show { state =>
+      val entries = state.asMap
+        .map { case (key, value) => s"$key=$value" }
+        .mkString("{", ",", "}")
+
+      s"TraceState{entries=$entries}"
+    }
+
+  /** Creates [[TraceState]] from the given map.
+    *
+    * '''Important''': the map entries will not be validated. Use this method
+    * when you 100% sure the entries are valid. For example, when you wrap
+    * OpenTelemetry Java trace state.
+    */
+  private[otel4s] def fromMapUnsafe(map: Map[String, String]): TraceState =
```

Comment:
this ought to use a `SeqMap`, as according to [the spec](https://www.w3.org/TR/trace-context/#mutating-the-tracestate-field):

> When mutating `tracestate`, the order of unmodified key/value pairs _MUST_ be preserved.

--------

unfortunately, there is an additional complication to contend with. the spec for the updating "mutation" states:

> The value for any given key can be updated. Modified keys _SHOULD_ be moved to the beginning (left) of the list.

as far as I can tell, none of the implementations of `SeqMap` do this (they keep the key in the same spot); additionally, I recall the scala team being very particular about behavioural consistency across implementations, though I don't seem to see the actual behaviour specified anywhere. and there is no additional functionality to change the behaviour (I believe it was suggested at some point and rejected).

--------

finally, there is another issue that's _surprisingly_ difficult to work around:

> The new key/value pair _SHOULD_ be added to the beginning of the list.

with the exception of `ListMap`, implementations put new entries at the _end_ of the `SeqMap`, and `SeqMap` has _no_ API changes over `Map`—that is, there is no way to get a constant-time `Seq` or `SeqView` out, and no `reverseIterator` (which would be constant time for `VectorMap` at least). I cannot believe I'm saying this, but `ListMap` may actually be our best option

Reply:
Good catch, thanks. I missed this part entirely. I've updated the implementation